R/batch-chat.R

```diff
@@ -0,0 +1,333 @@
+  )
+}
+
+BatchJob <- R6::R6Class(
+  "BatchJob",
+  public = list(
+    chat = NULL,
+    user_turns = NULL,
+    path = NULL,
+    should_wait = TRUE,
+    type = NULL,
+
+    # Internal state
+    provider = NULL,
+    started_at = NULL,
+    stage = NULL,
+    batch = NULL,
+    results = NULL,
+
+    initialize = function(
+      chat,
+      prompts,
+      path,
+      type = NULL,
+      wait = TRUE,
+      call = caller_env(2)
+    ) {
+      check_chat(chat, call = call)
+      self$provider <- chat$get_provider()
+      check_has_batch_support(self$provider, call = call)
+
+      user_turns <- as_user_turns(prompts, call = call)
+      check_string(path, allow_empty = FALSE, call = call)
+      check_bool(wait, call = call)
+
+      self$chat <- chat
+      self$user_turns <- user_turns
+      self$type <- type
+      self$path <- path
+      self$should_wait <- wait
+
+      if (file.exists(path)) {
+        state <- jsonlite::read_json(path)
+        self$stage <- state$stage
+        self$batch <- state$batch
+        self$results <- state$results
+        self$started_at <- .POSIXct(state$started_at)
+
+        self$check_hash(state$hash, call = call)
+      } else {
+        self$stage <- "submitting"
+        self$batch <- NULL
+        self$started_at <- Sys.time()
+      }
+    },
+
+    save_state = function() {
+      jsonlite::write_json(
+        list(
+          version = 1,
+          stage = self$stage,
+          batch = self$batch,
+          results = self$results,
+          started_at = as.integer(self$started_at),
+          hash = self$compute_hash()
+        ),
+        self$path,
+        auto_unbox = TRUE,
+        pretty = TRUE
+      )
+    },
+
+    step = function() {
+      if (self$stage == "submitting") {
+        self$submit()
+      } else if (self$stage == "waiting") {
+        self$wait()
+      } else if (self$stage == "retrieving") {
+        self$retrieve()
+      } else {
+        cli::cli_abort("Unknown stage: {self$stage}", .internal = TRUE)
+      }
+    },
+
+    step_until_done = function() {
+      while (self$stage != "done") {
+        if (!self$step()) {
+          return(invisible())
+        }
+      }
+      invisible(self)
+    },
+
+    submit = function() {
+      existing <- self$chat$get_turns(include_system_prompt = TRUE)
+      conversations <- append_turns(list(existing), self$user_turns)
+
+      self$batch <- batch_submit(self$provider, conversations, type = self$type)
+      self$stage <- "waiting"
+      self$save_state()
+      TRUE
+    },
+
+    wait = function() {
+      # always poll once, even when wait = FALSE
+      status <- self$poll()
+
+      if (self$should_wait) {
+        cli::cli_progress_bar(
+          format = paste(
+            "{cli::pb_spin} Processing... ",
+            "{status$n_processing} -> {cli::col_green({status$n_succeeded})} / {cli::col_red({status$n_failed})} ",
+            "[{self$elapsed()}]"
```

Comment:
```suggestion
            "{cli::pb_spin} Processing...",
            "[{self$elapsed()}]",
            "{status$n_processing} pending |", 
            "{cli::col_green({status$n_succeeded})} done |",
            "{cli::col_red({status$n_failed})} failed"
```

![image](https://github.com/user-attachments/assets/b22ba87d-797d-4314-adde-b7e64f08e474)

As a user, I think this format is much friendlier because I never have to guess about what the numbers mean. I do think they're _eventually_ clear, but it's nice to just not have to ask the question in the first place.